### PR TITLE
Pairing via Bluetooth with the RhythmActivity

### PR DESCRIPTION
**Pull request by clenrock**
- Relevant Issues: #109 #104 #61 
- Related Pull Requests: -

* * *

## What
This Pull Request adds to the repository the integrated version of the RhythmActivity together with the whole flow of how the pairing will work.

## Why
This Pull Request is needed because first it were only separate parts, but now it's connected together.

## How
This feature can be viewed/tested within the project by pairing over Bluetooth.

## Alternative implementation
Other implementations that I've have considered are 

## Notes
BIG NOTE: this is not fully tested, the PR is opened for now to review the code, but currently we are testing still and trying to find errors which we can get out.

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 1]
```diff
@@ -50,6 +48,7 @@
     private IBluetoothHandler bluetoothHandler;
     private Set<BluetoothDevice> pairedDevices;
     private ArrayAdapter<String> newDevicesArrayAdapter;
+    private boolean master = false;
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: Can you rename it to isMaster? Shift-F6 to rename a variable

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 2]
```diff
@@ -111,15 +114,26 @@ protected void onCreate(final Bundle savedInstanceState) {
...
+     */
+    @Override
     public void onStart() {
         super.onStart();
-
-        this.serviceLocator.registerToEventBus(this);
-        this.queryPairedDevices();
-        this.discoverDevices();
+        // Get the Paired Devices list
+        queryPairedDevices();
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: this

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 3]
```diff
@@ -207,10 +224,12 @@ public void onEvent(final BluetoothConnectedEvent event) {
     private void addNewDevice(final Intent intent) {
         final BluetoothDevice device = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);
 
-        // Skip paired devices and devices without a name.
-        if (device.getBondState() != BluetoothDevice.BOND_BONDED && !device.getName().equals("null")) {
-            this.newDevices.add(device);
-            this.newDevicesArrayAdapter.add(device.getName() + "\n" + device.getAddress());
+        if(device != null && device.getName() != null) {
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: Ctrl-alt-shift-r

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 4]
```diff
@@ -248,14 +252,43 @@ private void fillDatabaseWithDemoData() throws IOException {
     }
 
     /**
+     * Called when a Bluetooth connection is established.
+     *
+     * @param event Contains additional data about the event
+     */
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    public void onNewBluetoothConnectedEvent(final BluetoothConnectedEvent event) {
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Is this method needed?

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    /**
+     * Gets triggered when the start recording button is clicked.
+     *
+     * @param v - the {@link View} clicked
+     */
+    public void onStartRecordingClick(final View v) {
+        LOGGER.info("Start Recording clicked");
+        tapCombination = new ArrayList<>();
+
+        final Button startButton = (Button) this.findViewById(R.id.start_recording_button);
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Could you move these assignment to onCreate (still to final variables though) because you now have several times the same code (also in onSteopRecordingClick)

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                //Go to RhythmActivity
+                final Intent intent = new Intent(this, RhythmCreateActivity.class);
+                intent.putExtra(ConstantKeywords.SERVICE_LOCATOR, serviceLocator);
+                startActivity(intent);
+            } else if (verificationMessage.equals("visual")) {
+                //Go to VisualVerificationActivity
+                final Intent intent = new Intent(this, VisualVerificationActivity.class);
+                intent.putExtra(ConstantKeywords.SERVICE_LOCATOR, serviceLocator);
+                startActivity(intent);
+            }
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: else log it and show a SweetAlert that something went horribly wrong

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidBluetoothConnectThread.java [hunk 7]
```diff
@@ -83,10 +83,10 @@ public void run() {
      */
     void cancel() {
         LOGGER.warn("Cancelled!");
-        try {
-            socket.close();
-        } catch (final IOException e) {
-            LOGGER.error("Closing socket failed", e);
-        }
+//        try {
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidBluetoothConnectThread.java]: We'll fix this in issue #197

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidBluetoothConnectedThread.java [hunk 8]
```diff
@@ -65,10 +65,12 @@ public void run() {
         final byte[] buffer = new byte[BUFFER_SIZE];
         try {
             // Read from the InputStream
-            final int bytes = inStream.read(buffer);
-            LOGGER.info("Read {} bytes", bytes);
+            while (true) {
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidBluetoothConnectedThread.java]: Nice bug-fix!

File: app/src/main/res/layout/content_rhythm_create.xml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    <Button
+        android:id="@+id/done_tapping_button"
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
+        android:layout_weight="1"
+        android:background="@color/color_accent"
+        android:text="@string/done"
+        android:textColor="@color/white"
+        android:visibility="gone"
+        android:onClick="onDoneClick"/>
```
Comment (jverbraeken) [inline app/src/main/res/layout/content_rhythm_create.xml]: Can you make this method name a bit more descriptive? Again Shift-F6

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 4 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Yes, it's neat that you can also start from MainActivity

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 5 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Good point

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 10]
```diff
@@ -119,15 +123,25 @@ protected void onCreate(final Bundle savedInstanceState) {
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
     public void onStart() {
         super.onStart();
-
-        this.serviceLocator.registerToEventBus(this);
+        // Get the Paired Devices list
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: Unnecessary (the name of the method is `queryPairedDevices`) and possible confusing (`discoveredDevices` doesn't get paired devices`) comment

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 11]
```diff
@@ -119,15 +123,25 @@ protected void onCreate(final Bundle savedInstanceState) {
...
         this.queryPairedDevices();
         this.discoverDevices();
         LOGGER.info("Started the service and started discovering");
     }
 
+
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: While you're at it, can you remove this 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 12]
```diff
@@ -215,7 +247,8 @@ private void addNewDevice(final Intent intent) {
         final BluetoothDevice device = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);
 
         // Skip paired devices and devices without a name.
-        if (device.getBondState() != BluetoothDevice.BOND_BONDED && !device.getName().equals("null")) {
+        if (device != null && device.getName() != null
+            && device.getBondState() != BluetoothDevice.BOND_BONDED && !device.getName().equals("null")) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: Could you maybe refactor this into a separate method named `isDevicePaired` or something like that, this if-statement is getting a bit hard to read 😅

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 13]
```diff
@@ -246,6 +279,7 @@ public void onItemClick(final AdapterView<?> av, final View v, final int arg2, f
             // Get the device MAC address, which is the last 17 chars in the View
             final String info = ((TextView) v).getText().toString();
             if (!info.equals(getString(R.string.no_devices_found))) {
+                isMaster = true;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: Why is `isMaster` being set to true when the string equals to `R.string.no_devices_found`?

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 14]
```diff
@@ -97,6 +100,18 @@ protected void onCreate(final Bundle savedInstanceState) {
             LOGGER.info("Bluetooth handler not started, most likely Bluetooth is not enabled");
             this.enableBluetooth(false);
         }
+        if (intent.getSerializableExtra(ConstantKeywords.SUCCESSFUL_BLUETOOTH) != null) {
+
+            final boolean successfulBluetooth = (boolean) intent.getSerializableExtra(ConstantKeywords.SUCCESSFUL_BLUETOOTH);
+
+            if (successfulBluetooth) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Could you refactor this into a separate method? The `onCreate` method is getting a bit crowded 😄

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 15]
```diff
@@ -173,16 +189,18 @@ public void onPairingButtonClicked(final View view) {
                 break;
             case R.id.pairing_menu_nfc:
                 intent.setComponent(new ComponentName(this, NFCActivity.class));
+                intent.putExtra(ConstantKeywords.SERVICE_LOCATOR, this.serviceLocator);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Unnecessary use of `.putExtra`, see [line 177](https://github.com/ericcornelissen/NervousFish/pull/195/files#diff-149538cc1f1f5e4370b6ae93d6489330R177) of this file

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 16]
```diff
@@ -173,16 +189,18 @@ public void onPairingButtonClicked(final View view) {
...
+                this.startActivity(intent);
                 break;
             case R.id.pairing_menu_qr:
                 intent.setComponent(new ComponentName(this, QRExchangeKeyActivity.class));
+                intent.putExtra(ConstantKeywords.SERVICE_LOCATOR, this.serviceLocator);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Unnecessary use of `.putExtra`, see [line 177](https://github.com/ericcornelissen/NervousFish/pull/195/files#diff-149538cc1f1f5e4370b6ae93d6489330R177) of this file

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 17]
```diff
@@ -173,16 +189,18 @@ public void onPairingButtonClicked(final View view) {
...
+                this.startActivity(intent);
                 break;
             default:
                 LOGGER.error("Unknown pairing button clicked");
                 throw new IllegalArgumentException("Only existing buttons can be clicked");
         }
-
-        this.startActivity(intent);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Could you revert this change, the original code reduced duplicate code 😅

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 18]
```diff
@@ -225,14 +243,43 @@ private void fillDatabaseWithDemoData() throws IOException {
...
-        if (event.getClazz().equals(Contact.class)) {
+        if (event.getClazz().equals(String.class)) {
+            final String verificationMessage = (String) event.getData();
+
+            if ("rhythm".equals(verificationMessage)) {
+                //Go to RhythmActivity
+                final Intent intent = new Intent(this, RhythmCreateActivity.class);
+                intent.putExtra(ConstantKeywords.SERVICE_LOCATOR, serviceLocator);
+                this.startActivityForResult(intent, 0);
+            } else if ("visual".equals(verificationMessage)) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Could you change this if-statements from `"method".equals(verificationMessage)` to `verificationMessage.equals("method")` 😇 

Also, the comments `//Go to the ...Activity` seem a bit unnecessary to me...

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.List;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+/**
+ * The RhythmCreateActivity is an Activity where you can tap a sequence.
+ */
+//List is cast to an ArrayList, but that is needed to put in an intent
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Are you sure about this? Why can't a `List` be put into an intent? And since `ArrayList` is an extension of `List`, how is it possible that it can be put into an intent?

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+@SuppressFBWarnings(value = "BC_BAD_CAST_TO_CONCRETE_COLLECTION")
+public final class RhythmCreateActivity extends AppCompatActivity {
+    private static final Logger LOGGER = LoggerFactory.getLogger("RhythmCreateActivity");
+    private Button startButton;
+    private Button stopButton;
+    private Button doneButton;
+    private List<SingleTap> tapCombination;
+    private IServiceLocator serviceLocator;
+    //TODO: change contact into an encrypted string of bytes
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Could you remove this TODO and make an issue for it 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+    @Override
+    protected void onCreate(final Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_rhythm_create);
+        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar_create_rhythm);
+        setSupportActionBar(toolbar);
+
+        if (getSupportActionBar() != null) {
+            getSupportActionBar().setDisplayShowTitleEnabled(false);
+        }
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Is this if-statement needed?

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+     * @param v - the {@link View} clicked
+     */
+    public void onStartRecordingClick(final View v) {
+        LOGGER.info("Start Recording clicked");
+        tapCombination = new ArrayList<>();
+
+        startButton.setVisibility(View.GONE);
+
+        stopButton.setVisibility(View.VISIBLE);
+
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: While you're at it, could you remove the excessive enters? 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+     *
+     * @param v - the {@link View} clicked
+     */
+    public void onStopRecordingClick(final View v) {
+        LOGGER.info("Stop Recording clicked");
+
+        startButton.setVisibility(View.VISIBLE);
+
+        stopButton.setVisibility(View.GONE);
+
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: While you're at it, could you remove the excessive enters? 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void onStart() {
+        super.onStart();
+        this.serviceLocator.registerToEventBus(this);
+    }
+
+    @Override
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Missing JavaDoc

File: app/src/main/java/com/nervousfish/nervousfish/activities/SelectVerificationMethodActivity.java [hunk 25]
```diff
@@ -53,11 +55,21 @@ public void openVerificationMethod(final View view) {
             case R.id.select_visual_verification:
                 LOGGER.info("Selected visual verification method, opening activity");
                 intent.setComponent(new ComponentName(this, VisualVerificationActivity.class));
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SelectVerificationMethodActivity.java]: Can you move the `.setComponent` to after the try-catch block, similar to the `R.id.select_rhythm_verification` case 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java [hunk 26]
```diff
@@ -70,4 +93,38 @@ public void buttonAction(final View v) {
...
     }
 
+    @Override
+    protected void onStart() {
+        super.onStart();
+        this.serviceLocator.registerToEventBus(this);
+    }
+
+    @Override
+    protected void onStop() {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Missing JavaDoc

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java [hunk 27]
```diff
@@ -70,4 +93,38 @@ public void buttonAction(final View v) {
         }
     }
 
+    @Override
+    protected void onStart() {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Missing JavaDoc

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+import org.greenrobot.eventbus.Subscribe;
+import org.greenrobot.eventbus.ThreadMode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Used to let the Bluetooth-initiating user know that he should wait for his partner
+ * to complete the pairing session.
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Isn't this class used in a broader sense then is being described here?

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+
+    private void evaluateData() {
+        //TODO: check if when we decrypt the dataReceived with the tapCombination that we get a normal contact
+        LOGGER.info("Evaluating data");
+        final Intent intent = new Intent(this, MainActivity.class);
+        intent.putExtra(ConstantKeywords.SERVICE_LOCATOR, serviceLocator);
+        intent.putExtra(ConstantKeywords.SUCCESSFUL_BLUETOOTH, true);
+        this.startActivity(intent);
+        //setResult(ConstantKeywords.DONE_PAIRING_RESULT_CODE);
+        //finish();
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Could you remove commented code 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+
+        final String message = (String) intent.getSerializableExtra(ConstantKeywords.WAIT_MESSAGE);
+        final TextView waitingMessage = (TextView) findViewById(R.id.waiting_message);
+        waitingMessage.setText(message);
+
+        LOGGER.info("WaitActivity created");
+    }
+
+    private void evaluateData() {
+        //TODO: check if when we decrypt the dataReceived with the tapCombination that we get a normal contact
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Can you remove this `TODO:` comment and include it in the issue I requested you to make 😄

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void onStart() {
+        super.onStart();
+        if (this.dataReceived != null && tapCombination != null) {
+            evaluateData();
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Isn't `this.dataReceived != null && tapCombination != null` part of evaluating the data?

File: app/src/main/res/drawable-hdpi/ripple_effect.xml [hunk 32]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (ericcornelissen) [inline app/src/main/res/drawable-hdpi/ripple_effect.xml]: Why did you add this file to `~/drawable-hdpi/`?

File: config/quality/checkstyle/checkstyle.xml [hunk 33]
```diff
@@ -50,7 +50,7 @@
         <module name="EmptyStatement" />
         <module name="EqualsAvoidNull" />
         <module name="EqualsHashCode" />
-        <module name="ExplicitInitialization">
+        <module name="ExplicitInitialization" >
```
Comment (ericcornelissen) [inline config/quality/checkstyle/checkstyle.xml]: Can you revert this change 😇

File: app/src/main/res/values-nl/strings.xml [hunk 34]
```diff
@@ -92,4 +92,13 @@
     <string name="contact_set_name">\"Geef contact een naam\"</string>
     <string name="contact_standard_new">Nieuwe contact</string>
     <string name="popup_done">Klaar</string>
+    <string name="wait_message_slave_verification_method">Wachten tot het andere apparaat een verificatie methode heeft gekozen...</string>
```
Comment (ericcornelissen) [inline app/src/main/res/values-nl/strings.xml]: Could you organize these strings a bit better?

File: app/src/main/res/values/strings.xml [hunk 35]
```diff
@@ -70,6 +70,19 @@
     <string name="contact_deleted_description">The contact has been deleted!</string>
     <string name="contact_doesnt_exist">The contact doesn\'t exist</string>
     <string name="contact_already_deleted">It looks like the contact was already deleted.</string>
+
+    <string name="contact_added_popup_title">Pairing was successful!</string>
```
Comment (ericcornelissen) [inline app/src/main/res/values/strings.xml]: Could you organize these strings a bit better?

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 36]
```diff
@@ -29,17 +28,17 @@
...
 
 /**
  * An {@link Activity} that establishes and manages a bluetooth connection.
  * It shows a screen with the Bluetooth devices with which the user is paired and other Bluetooth devices
  * that are detected near the device of the user.
  */
-@SuppressWarnings("PMD.ExcessiveImports") // Uses many Android and utility classes
+@SuppressWarnings({"PMD.ExcessiveImports", "PMD.TooManyMethods"})
+// 1. Uses many Android and utility classes
+// 2. The amount of methods is not too much at this moment
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: Maybe change tge default value in the rules then?

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 37]
```diff
@@ -194,15 +223,18 @@ public void stopDiscovering() {
...
         LOGGER.info("onBluetoothConnectedEvent called");
-        try {
-            final List<Contact> list = this.serviceLocator.getDatabase().getAllContacts();
-            for (final Contact e : list) {
-                this.bluetoothHandler.send(e);
-            }
-        } catch (final IOException e) {
-            LOGGER.warn("Writing all contacts issued an IOexception", e);
+        if (isMaster) {
+            isMaster = false;
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: It shouldn't matter much but placing this pine at the end is peobably safer

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java (hunk 12 above)
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: I agree that there are a few null checks but everything verbose enough to me to make it clear. The new metgod would be so small and we already have a pmd warnibg about that.

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java (hunk 26 above)
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Why doesn't checkstyle trigger here so we don't have to worry about smaal things like this?

File: app/src/main/res/values-nl/strings.xml [hunk 38]
```diff
@@ -92,4 +92,13 @@
...
+    <string name="title_activity_rhythm_create">RhythmCreateActivity</string>
+    <string name="tap_me">Klik me!</string>
```
Comment (TheBonheurs) [inline app/src/main/res/values-nl/strings.xml]: Is it just me or does the dutch version sound awkward?

File: app/src/main/res/values/colors.xml [hunk 39]
```diff
@@ -6,7 +6,9 @@
 
     <color name="transparent_white">#33FFFFFF</color>
     <color name="red_alert">#9A3334</color>
+    <color name="ok_green">#009900</color>
 
-    <color name="red_fail">#dd6b55</color>
+    <color name="white">#fff</color>
```
Comment (TheBonheurs) [inline app/src/main/res/values/colors.xml]: Why this order change?

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java (hunk 26 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Not sure, have not yet looked at this issue 😅

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java (hunk 12 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: Implemented this suggestion to hopefully make it clearer

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java (hunk 13 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: The if statement is used to check if the name is not equals "No devices found..." because that string uses the same onclick method. I will refactor this to make it clearer.

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 17 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: No this was on purpose. Case "R.id.pairing_menu_bluetooth:" was changed to give a popup when Bluetooth was not enabled. After this is enabled via the popup it will go to the desired activity.

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 18 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Checkstyle told me to do it like this, do you want me to suppress this warning?

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 19 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: I'm not sure why the List cant be put in the intent. I searched internet but everyone said that you should use the implementation of List instead of the super class.

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 22 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: I used this because it gives much more readability, do you still want me to change this?

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java (hunk 12 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: @TheBonheurs I'm not sure, but I think that PMD error does not complain about private methods (at least I don't think it should)

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java (hunk 31 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: No, the data should only be evaluated when both are not null. Else we need to wait until the data is received via an Event and then evaluate it.
(You didn't even notice that I forgot "this." in the same line)

File: app/src/main/res/drawable-hdpi/ripple_effect.xml (hunk 32 above)
Comment (clenrock) [inline app/src/main/res/drawable-hdpi/ripple_effect.xml]: I'll add it to just 'drawable'

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 17 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: You can return early after displaying the popup (which is already the case, see [line 187](https://github.com/ericcornelissen/NervousFish/pull/195/files/8b94f909f6d413b81431fd77d2182ed44556b53c#diff-149538cc1f1f5e4370b6ae93d6489330R187)) 😉

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 18 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Checkstyle wants it like `"method".equals(verificationMessage)`!? Welp, just leave it like this then 😓

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 19 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Well, in that case maybe only accept `ArrayList`s in order to prevent any unforeseen issues?

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 22 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: I don't think it adds any readability since it concerns single lines. If you want to make clear these are discreet steps in the program, consider adding comments to make the distinction clear

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java (hunk 36 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: I think this depends on the largeness of the methods. This class has a lot of small methods where it is no problem, but somewhere else it could be better to have this PMD error.

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java (hunk 37 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: Why?

File: app/src/main/res/values-nl/strings.xml (hunk 38 above)
Comment (clenrock) [inline app/src/main/res/values-nl/strings.xml]: Hahaha, what would you suggest? Maybe "Klik hier" or only "Klik"?

Comment (TheBonheurs) [inline app/src/main/res/values-nl/strings.xml]: Klik hier imo

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java (hunk 37 above)
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: The master flag isnt being reset immediatly after checking it

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java (hunk 36 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: I agree with Cornel, generally you do not want to have many method In a class. Only in specific cases where the amount of methods cannot be reduced without cluttering the code this error should be suppressed

File: app/src/main/res/values-nl/strings.xml (hunk 38 above)
Comment (jverbraeken) [inline app/src/main/res/values-nl/strings.xml]: Tik hier :)

Comment (ericcornelissen) [inline app/src/main/res/values-nl/strings.xml]: I agree with "Tik hier" 👍

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 14 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: This is done, weird that it doesn't close the comment

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 17 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Good point!

File: app/src/main/res/values-nl/strings.xml (hunk 38 above)
Comment (clenrock) [inline app/src/main/res/values-nl/strings.xml]: Thanks for your input :)

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 14 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: [Are you sure?](https://github.com/ericcornelissen/NervousFish/blob/8b94f909f6d413b81431fd77d2182ed44556b53c/app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java#L107)

Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Yes, still weird that Github doesn't show it. When you check out on the branch this is put in a seperate method.

Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Oke, wierd 😕 can you replace the argument of that method (currently an `Intent`) with the result of `intent.getSerializableExtra(ConstantKeywords.SUCCESSFUL_BLUETOOTH)`
